build.assets/tooling/cmd/resource-ref-generator/reference/reference.go

```diff
@@ -0,0 +1,306 @@
+	}
+
+	switch {
+	case c.DestinationDirectory == "":
+		return errors.New("no destination path provided")
+	case c.FieldAssignmentMethodName == "":
+		return errors.New("must provide a field assignment method name")
+	case len(c.RequiredFieldTypes) == 0:
+		return errors.New("must provide a list of required field types")
+	case c.SourcePath == "":
+		return errors.New("must provide a source path")
+	default:
+		return nil
+	}
+}
+
+// getPackageInfoFromExpr extracts a package name and declaration name from an
+// arbitrary expression. If the expression is not an expected kind,
+// getPackageInfoFromExpr returns an empty PackageInfo.
+func getPackageInfoFromExpr(expr ast.Expr) resource.PackageInfo {
+	var gopkg, fldname string
+	switch t := expr.(type) {
+	case *ast.StarExpr:
+		return getPackageInfoFromExpr(t.X)
+	case *ast.SelectorExpr:
+		// If the type of the field is an *ast.SelectorExpr,
+		// it's of the form <package>.<type name>.
+		g, ok := t.X.(*ast.Ident)
+		if ok {
+			gopkg = g.Name
+		}
+		fldname = t.Sel.Name
+
+	// There's no package, so only assign a name.
+	case *ast.Ident:
+		fldname = t.Name
+	}
+	return resource.PackageInfo{
+		DeclName:    fldname,
+		PackageName: gopkg,
+	}
+}
+
+// shouldProcess indicates whether we should generate reference entries from the
+// type declaration represented in d (i.e., whether this is a dynamic resource
+// type). To do so, it checks whether d:
+//   - is a struct type
+//   - has fields with the required types
+//   - does not belong to the list of excluded resources
+func shouldProcess(d resource.DeclarationInfo, requiredTypes, excludedResources []TypeInfo) bool {
+	// We expect the declaration to be a type declaration with one spec.
+	gendecl, ok := d.Decl.(*ast.GenDecl)
+	if !ok {
+		return false
+	}
+
+	if len(gendecl.Specs) != 1 {
+		return false
+	}
+
+	t, ok := gendecl.Specs[0].(*ast.TypeSpec)
+	if !ok {
+		return false
+	}
+
+	if t == nil {
+		return false
+	}
+
+	// If the declaration type is not a struct, we can't process it as a
+	// root resource entry.
+	str, ok := t.Type.(*ast.StructType)
+	if !ok {
+		return false
+	}
+
+	// If the configured excluded resources include this type declaration,
+	// don't process it.
+	for _, r := range excludedResources {
+		if t.Name.Name == r.Name && d.PackageName == r.Package {
+			return false
+		}
+	}
+	// Use only the final segment of each desired package path
+	// in the comparison, since that is what is preserved in the
+	// AST.
+	finalTypes := make([]TypeInfo, len(requiredTypes))
+	for i, t := range requiredTypes {
+		segs := strings.Split(t.Package, "/")
+		finalTypes[i] = TypeInfo{
+			Package: segs[len(segs)-1],
+			Name:    t.Name,
+		}
+	}
+
+	// Compare the types of fields in the struct type with the required
+	// fields types. Only one required field type must be present.
+	var m bool
+	for _, fld := range str.Fields.List {
+		if len(fld.Names) != 1 {
+			continue
+		}
```

Comment:
This sill skip over things like:

```
type Foo struct {
    Bar, Baz int
}
```

Again, it's probably a safe assumption to make, but I wonder if we should at least log/warn about things that don't match our expectations rather than treating them like they don't exist.